### PR TITLE
scheduler: fix TestIncomingPodsMetrics unit test

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

addUnschedulablePodBackToBackoffQ happened to put the pod into the backoff queue because
- the pod was not popped earlier and thus not in flight
- the PodInfo had UnschedulablePlugins set
- determineSchedulingHintForInFlightPod has code for "if UnschedulablePlugins is set and pod not in flight -> internal error, use backoff"

Relying on such special code is not good. A better way to force backoff is by recording some concurrent event. isPodWorthRequeuing then calls the queueHintReturnQueueAfterBackoff function and the pod goes to the backoff queue.

#### Which issue(s) this PR fixes:
Related-to: https://github.com/kubernetes/kubernetes/pull/120413#discussion_r1315165907

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 1]
```diff
@@ -2033,17 +2033,34 @@ var (
...
+	expectNoError = func(what string, err error) {
+		if err != nil {
+			panic(fmt.Sprintf("Unexpected error from %s: %v", what, err))
+		}
+	}
 	addUnschedulablePodBackToUnschedulablePods = func(logger klog.Logger, queue *PriorityQueue, pInfo *framework.QueuedPodInfo) {
 		// To simulate the pod is failed in scheduling in the real world, Pop() the pod from activeQ before AddUnschedulableIfNotPresent() below.
-		queue.activeQ.Add(queue.newQueuedPodInfo(pInfo.Pod))
-		if p, err := queue.Pop(); err != nil || p.Pod != pInfo.Pod {
+		expectNoError("Add", queue.activeQ.Add(queue.newQueuedPodInfo(pInfo.Pod)))
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:2043]: https://github.com/kubernetes/community/blob/master/sig-scheduling/CONTRIBUTING.md#technical-and-style-guidelines

Use `t.Fatal`

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 2]
```diff
@@ -2033,17 +2033,34 @@ var (
...
+		p, err := queue.Pop()
+		expectNoError("Pop", err)
+		if p.Pod != pInfo.Pod {
 			panic(fmt.Sprintf("Expected: %v after Pop, but got: %v", pInfo.Pod.Name, p.Pod.Name))
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:2047]: uhm... this one slipped, please use t.Fatal

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 3]
```diff
@@ -2033,17 +2033,34 @@ var (
 	add = func(logger klog.Logger, queue *PriorityQueue, pInfo *framework.QueuedPodInfo) {
 		queue.Add(logger, pInfo.Pod)
 	}
+	expectNoError = func(what string, err error) {
+		if err != nil {
+			panic(fmt.Sprintf("Unexpected error from %s: %v", what, err))
+		}
+	}
 	addUnschedulablePodBackToUnschedulablePods = func(logger klog.Logger, queue *PriorityQueue, pInfo *framework.QueuedPodInfo) {
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:2041]: let's change the function name to: `popAndRequeueAsUnschedulable`. Similarly for the other one

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 4]
```diff
@@ -2033,17 +2033,34 @@ var (
...
+		// To simulate the pod is failed in scheduling in the real world, Pop() the pod from activeQ before AddUnschedulableIfNotPresent() below.
+		expectNoError("Add", queue.activeQ.Add(queue.newQueuedPodInfo(pInfo.Pod)))
+		p, err := queue.Pop()
+		expectNoError("Pop", err)
+		if p.Pod != pInfo.Pod {
+			panic(fmt.Sprintf("Expected: %v after Pop, but got: %v", pInfo.Pod.Name, p.Pod.Name))
+		}
+
+		// A concurrent event forces the pod into the backoff queue instead of the unschedulable queue.
+		queue.MoveAllToActiveOrBackoffQueue(logger, NodeAdd, nil, nil, nil)
```
Comment (alculquicondor) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:2061]: I don't like the fact that, to understand what each test case is doing, I have to go back and read this function.
I would prefer if we would just have:

operations: []{pop, moveAll, addUnschedulable, ...}

But up to you if you want to do it or leave it for someone else to follow up.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 1 above)
Comment (pohly) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:2043]: I would have preferred that myself. But `t` is not passed down into these operations. I could change that, but then the diff becomes larger.

Comment (pohly) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:2043]: Or did you mean not having a `expectNoError` function ([DAMP](https://stackoverflow.com/questions/6453235/what-does-damp-not-dry-mean-when-talking-about-unit-tests/11837973#11837973))?

I find `expectNoError` more readable than an if check. But if that's just me, then I can change to `if ... t.Fatal`.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 3 above)
Comment (pohly) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:2041]: Done.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 4 above)
Comment (pohly) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:2061]: Someone else, please :sweat:

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 2 above)
Comment (pohly) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:2047]: Done.

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 1 above)
Comment (pohly) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:2043]: Both done (using `t.Fatal` and removal of `expectNoError`).

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go]: Thanks! More importantly I wanted you to use `t.Fatal`.
But I also don't like adding "assert" libraries.

**Comment by pohly** [comment]
The tests pass in their branch, but not when merged into master. That's because of 0d3eafdfa39b23f035e66c4575490ed14a12c198: together with that, some pods in the unit tests no longer go to the expected queues:
```
   --- FAIL: TestIncomingPodsMetrics/add_pods_to_unschedulablePods_and_then_move_all_to_activeQ (0.10s)
        scheduling_queue.go:617: I0912 07:53:19.010475] Checking events for in-flight pod pod="ns1/test-pod-1" unschedulablePlugins={} inFlightEventsSize=1 inFlightPodsSize=1
        scheduling_queue.go:752: I0912 07:53:19.010547] Pod moved to an internal scheduling queue pod="ns1/test-pod-1" event="ScheduleAttemptFailure" queue="Backoff" schedulingCycle=1 hint="QueueAfterBackoff"
        scheduling_queue.go:617: I0912 07:53:19.010586] Checking events for in-flight pod pod="ns2/test-pod-2" unschedulablePlugins={} inFlightEventsSize=1 inFlightPodsSize=1
        scheduling_queue.go:752: I0912 07:53:19.010620] Pod moved to an internal scheduling queue pod="ns2/test-pod-2" event="ScheduleAttemptFailure" queue="Backoff" schedulingCycle=1 hint="QueueAfterBackoff"
        scheduling_queue.go:617: I0912 07:53:19.010652] Checking events for in-flight pod pod="ns3/test-pod-3" unschedulablePlugins={} inFlightEventsSize=1 inFlightPodsSize=1
        scheduling_queue.go:752: I0912 07:53:19.010681] Pod moved to an internal scheduling queue pod="ns3/test-pod-3" event="ScheduleAttemptFailure" queue="Backoff" schedulingCycle=1 hint="QueueAfterBackoff"
        scheduling_queue_test.go:2939: unexpected collecting result:
            
            
            Diff:
            --- metric output does not match expectation; want
            +++ got:
            @@ -2,4 +2,3 @@
             # TYPE scheduler_queue_incoming_pods_total counter
            -scheduler_queue_incoming_pods_total{event="ScheduleAttemptFailure",queue="unschedulable"} 3
            -scheduler_queue_incoming_pods_total{event="UnschedulableTimeout",queue="active"} 3
            +scheduler_queue_incoming_pods_total{event="ScheduleAttemptFailure",queue="backoff"} 3
```

I'll rebase and check whether the unit tests made some invalid assumptions about how the queue works.

**Comment by pohly** [comment]
The relevant difference is that pods with no unschedulable plugins now go into the backoff queue. To simulate putting a pod into the unschedulable queue, `pInfo.UnschedulablePlugins` must be set. Because of this change of behavior, a concurrent event is no longer needed when the backoff queue is the intended target:

```patch
diff --git a/pkg/scheduler/internal/queue/scheduling_queue_test.go b/pkg/scheduler/internal/queue/scheduling_queue_test.go
index 13a9a89bb02..b17e5a3c3cc 100644
--- a/pkg/scheduler/internal/queue/scheduling_queue_test.go
+++ b/pkg/scheduler/internal/queue/scheduling_queue_test.go
@@ -2288,6 +2288,8 @@ var (
        }
        popAndRequeueAsUnschedulable = func(t *testing.T, logger klog.Logger, queue *PriorityQueue, pInfo *framework.QueuedPodInfo) {
                // To simulate the pod is failed in scheduling in the real world, Pop() the pod from activeQ before AddUnschedulableIfNotPresent() below.
+               // UnschedulablePlugins will get cleared by Pop, so make a copy first.
+               unschedulablePlugins := pInfo.UnschedulablePlugins.Clone()
                if err := queue.activeQ.Add(queue.newQueuedPodInfo(pInfo.Pod)); err != nil {
                        t.Fatalf("Unexpected error during Add: %v", err)
                }
@@ -2298,6 +2300,8 @@ var (
                if p.Pod != pInfo.Pod {
                        t.Fatalf("Expected: %v after Pop, but got: %v", pInfo.Pod.Name, p.Pod.Name)
                }
+               // Simulate plugins that are waiting for some events.
+               p.UnschedulablePlugins = unschedulablePlugins
                if err := queue.AddUnschedulableIfNotPresent(logger, p, 1); err != nil {
                        t.Fatalf("Unexpected error during AddUnschedulableIfNotPresent: %v", err)
                }
@@ -2314,10 +2318,7 @@ var (
                if p.Pod != pInfo.Pod {
                        t.Fatalf("Expected: %v after Pop, but got: %v", pInfo.Pod.Name, p.Pod.Name)
                }
-
-               // A concurrent event forces the pod into the backoff queue instead of the unschedulable queue.
-               queue.MoveAllToActiveOrBackoffQueue(logger, NodeAdd, nil, nil, nil)
-
+               // When there is no known unschedulable plugin, pods always go to the backoff queue.
                if err := queue.AddUnschedulableIfNotPresent(logger, p, 1); err != nil {
                        t.Fatalf("Unexpected error during AddUnschedulableIfNotPresent: %v", err)
                }
```

**Review by sanposhiho — APPROVED** [review]
/lgtm
/approve

**Pull request merged**